cpp/src/arrow/flight/sql/odbc/flight_sql/utils.cc

```diff
@@ -0,0 +1,1124 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/flight/sql/odbc/flight_sql/utils.h"
+
+#include "arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/calendar_utils.h"
+#include "arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/encoding.h"
+#include "arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/platform.h"
+#include "arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/types.h"
+
+#include "arrow/builder.h"
+#include "arrow/compute/api.h"
+#include "arrow/type.h"
+#include "arrow/type_fwd.h"
+
+#include "arrow/flight/sql/odbc/flight_sql/json_converter.h"
+
+#include <boost/tokenizer.hpp>
+
+#include <ctime>
+#include <sstream>
+
+namespace driver {
+namespace flight_sql {
+
+namespace {
+bool IsComplexType(arrow::Type::type type_id) {
```

Comment:
The codebase usually calls these 'nested' types. Also, there's already `is_nested` in type_traits.h